internal/pkg/cli/deploy/svc.go

```diff
@@ -241,10 +242,18 @@ func newWorkloadDeployer(in *WorkloadDeployerInput) (*workloadDeployer, error) {
 	if err != nil {
 		return nil, fmt.Errorf("get application %s resources from region %s: %w", in.App.Name, in.Env.Region, err)
 	}
-	addonsSvc, err := addon.New(in.Name)
+
+	s3Client := s3.New(envSession)
+	var addonsSvc stackBuilder
+	if in.UploadAddonAssets {
+		addonsSvc, err = addon.NewPackager(in.Name, resources.S3Bucket, s3Client)
+	} else {
+		addonsSvc, err = addon.New(in.Name)
+	}
```

Comment:
💭 💭 Feel free to ignore. I just want to share my thoughts!

What if we had
```go
package addon

// Template now returns the *Template instead of string.
func (a *Addon) Template() (*Template, error)

func (tpl *Template) String() string

func (tpl *Template) Package(s3 Uploader, bucket string) error
```

and then we could have 
```go
package cli/deploy

func (d *workloadDeployer) pushAddonsTemplateToS3Bucket(in *pushAddonsTemplateToS3BucketInput) (string, error)
     tmpl, err := d.addons.Template()
     err := tmpl.Package(in.uploader, d.resources.S3Bucket)
     reader := strings.NewReader(tmpl.String())
     url, err := in.uploader.Upload(d.resources.S3Bucket, artifactpath.Addons(d.name, []byte(tmpl.String())), reader)
     return url, nil
}
```
This decouples the `pkg` step from `read` step and `swap` step, and lets the caller decide when to `pkg`, and when to `swap`. 

What poked me into seeking for an alternative was the fact that we are implicitly setting how `addons` behave by using different constructors. Here, we use different constructors to initialize an `addons` because  in some downstream calls we want `addons` to/not to package assets. In the long run, this may become difficult to debug, because for example, to understand what a method call does we'd need to trace all the way to when the receiver was constructed. I do believe there are other corners in this codebase that do the same thing - as long as we can keep track, it's no biggie. 

Just some food for thoughts! The current approach also LGTM!

Reply:
i think this sounds great!! i'm going to leave it as is for now - and get all of the functionality merged, and then revisit this and make it cleaner 😊